### PR TITLE
fix: prevent front-running attack in GenericHarvester

### PR DESCRIPTION
**Pull request by Joh077**
- Add leftover token detection in harvest() to protect against router manipulation
- Revert if router doesn't consume all tokens received from Parallelizer
- Add RouterDidNotConsumeAllTokens custom error
- Add tests 

Fix compares token balance before receiving from Parallelizer and after router swap. Any leftover triggers revert, protecting user funds on flash loan

### CODE REVIEW

File: tests/units/GenericHarvester.t.sol [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    require(
+      receiver.onFlashLoan(msg.sender, _token, amount, 0, data) == keccak256("ERC3156FlashBorrower.onFlashLoan"),
+      "Callback failed"
+    );
+
+    token.transferFrom(address(receiver), address(this), amount);
+
+    return true;
+  }
+}
```
Comment (FabienCoutant) [inline tests/units/GenericHarvester.t.sol]: Mock should be move into mock/ folder
Also MockRouter/MockOdosRouter already exist

File: tests/units/GenericHarvester.t.sol [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+    uint256 tokenInBalanceFinal = IERC20(tokenIn).balanceOf(address(this));
+    if (tokenInBalanceFinal > tokenInBalanceBefore) {
+      revert RouterDidNotConsumeAllTokens();
+    }
+  }
+}
+
+/// @title GenericHarvester Front-Run Protection Tests
+/// @notice Tests for the leftover token detection mechanism that protects against front-running attacks
+contract Test_GenericHarvester_FrontRunProtection is Fixture {
```
Comment (FabienCoutant) [inline tests/units/GenericHarvester.t.sol]: no need to put FrontRunProctection in name, you can test the all contract here not just the FrontRunProctection

File: tests/units/GenericRebalancer.t.sol [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+
+    // Harvest should revert when malicious router leaves leftover tokens
+    vm.expectRevert(RouterDidNotConsumeAllTokens.selector);
+    vm.prank(alice);
+    harvester.harvest(address(eurB), 5e8, extraData);
+
+    // User budget should remain unchanged (protected from loss)
+    assertEq(harvester.budget(alice), 10 ether);
+  }
+}
```
Comment (FabienCoutant) [inline tests/units/GenericRebalancer.t.sol]: Why do you have integrations test under tests/units ?

File: tests/units/GenericHarvester.t.sol [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    IERC20(tokenIn).approve(router, amountToSwap);
+
+    MockRouterWithLeftovers(router).swap(amountToSwap, tokenIn, amountOut, tokenOut);
+
+    uint256 tokenInBalanceFinal = IERC20(tokenIn).balanceOf(address(this));
+    if (tokenInBalanceFinal > tokenInBalanceBefore) {
+      revert RouterDidNotConsumeAllTokens();
+    }
+  }
+}
```
Comment (FabienCoutant) [inline tests/units/GenericHarvester.t.sol]: What is the point of this contract?

File: tests/units/GenericHarvester.t.sol [hunk 5]
```diff
@@ -0,0 +1,233 @@
...
+    MockParallelizer mockParallelizer = new MockParallelizer();
+    LeftoverDetector detector = new LeftoverDetector();
+
+    deal(address(eurA), address(mockParallelizer), 100 ether);
+    deal(address(eurB), address(maliciousRouter), 50 ether);
+
+    vm.expectRevert(RouterDidNotConsumeAllTokens.selector);
+    detector.swapWithLeftoverDetection(
+      address(eurA), address(eurB), address(mockParallelizer), address(maliciousRouter), 100 ether, 100 ether, 50 ether
+    );
```
Comment (FabienCoutant) [inline tests/units/GenericHarvester.t.sol]: I should see a genereicHarvester.harvest(...) with most real scenario.

File: tests/units/GenericHarvester.t.sol (hunk 1 above)
Comment (Joh077) [reply tests/units/GenericHarvester.t.sol]: `MockRouterWithLeftovers.sol`& `MockFlashLoan.sol` moved to `tests/mock/`
For MockRouter/MockOdosRouter, I considered using the existing MockOdosRouter, but it doesn't support our test requirements

MockOdosRouter consumes exactly amountIn parameter (fixed amount) and i needed configurable partial consumption (70%, 99.9%, etc.) to simulate malicious routers
MockRouterWithLeftovers allows to test the leftover detection with different consumption rates without modifying existing mocks (which could break other tests).

File: tests/units/GenericHarvester.t.sol (hunk 2 above)
Comment (Joh077) [reply tests/units/GenericHarvester.t.sol]: Renamed to `Test_GenericHarvester`for units and `Test_GenericHarvester_Integration` for integrations

File: tests/units/GenericRebalancer.t.sol (hunk 3 above)
Comment (Joh077) [reply tests/units/GenericRebalancer.t.sol]: Fixed, I've moved integration tests to tests/integrations/GenericHarvester.t.sol. Now they're properly separate

File: tests/units/GenericHarvester.t.sol (hunk 4 above)
Comment (Joh077) [reply tests/units/GenericHarvester.t.sol]: This was my first approach to test the leftover detection separately, but after your comment I refactored everything to use the real harvester.harvest() instead. Way better for testing the actual production flow

File: tests/units/GenericHarvester.t.sol (hunk 5 above)
Comment (Joh077) [reply tests/units/GenericHarvester.t.sol]: Done! Refactored everything to use harvester.harvest() with real scenarios, much better approach. I also added a bunch of tests while refactoring, leftover detection at different rates (1%, 50%, 70%, 90%, 99.9%, 100%), scale validation (zero, partial, max, exceeds max), access control checks, budget management edge cases and state verification on reverts 

Ended up with 18 unit tests + 3 integration tests

**Comment by Joh077** [comment]
Update :  fix vulnerability + test coverage

fix(GenericRebalancer): credit budget surplus when amountStableOut exceeds flashloan amount (missing else if)
test(GenericRebalancer): new unit tests (flash loan guards, increase direction, budget underflow) + integration (4 end-to-end scenarios)
fix: lint & ci issues

File: contracts/helpers/GenericRebalancer.sol [hunk 6]
```diff
@@ -153,17 +153,26 @@ contract GenericHarvester is BaseHarvester, IERC3156FlashBorrower, RouterSwapper
         tokenOut = asset;
       }
     }
+
+    uint256 tokenInBalanceBefore = IERC20(tokenIn).balanceOf(address(this));
+
     uint256 amountOut =
       parallelizer.swapExactInput(amount, 0, address(tokenP), tokenIn, address(this), block.timestamp);
 
     // Swap to tokenIn
```
Comment (FabienCoutant) [inline contracts/helpers/GenericRebalancer.sol]: wrong comment: swap to tokenOut

File: contracts/helpers/GenericRebalancer.sol [hunk 7]
```diff
@@ -153,17 +153,26 @@ contract GenericHarvester is BaseHarvester, IERC3156FlashBorrower, RouterSwapper
...
     amountOut = _swapToTokenOut(typeAction, tokenIn, tokenOut, amountOut, swapType, callData);
 
+    //  Ensure router consumed all tokens
+    if (IERC20(tokenIn).balanceOf(address(this)) > tokenInBalanceBefore) {
+      revert RouterDidNotConsumeAllTokens();
+    }
```
Comment (FabienCoutant) [inline contracts/helpers/GenericRebalancer.sol:168]: Will always revert in case of swapExactOutput with positive slippage

**Review by FabienCoutant — APPROVED** [review]
